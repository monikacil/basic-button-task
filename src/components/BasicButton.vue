<template>
  <component class="btn" :is="type" :class="btnStyle" :href="href" :to="to" :disabled="disabled">
    <slot />
  </component>
</template>

<script setup lang="ts">
import { computed } from 'vue'

interface Props {
  color: string
  href?: string
  to?: string
  size?: string
  variant?: string
  disabled?: boolean
  fullWidth?: boolean
}

/**
 * Possible value for dynamic props:
 * color (required): orange | green | dark-green | yellow | red
 * size: s | m (default) | l
 * variant: filled (default) | outlined
 *
 */

// define props with default value
const props = withDefaults(defineProps<Props>(), {
  size: 'm',
  variant: 'filled',
  disabled: false,
  fullWidth: false
})

// computed button type
const type = computed(() => {
  if (props.href) {
    return 'a'
  } else if (props.to) {
    return 'router-link'
  } else {
    return 'button'
  }
})

// computed button styles based on given props
const btnStyle = computed<string[]>(() => {
  const styles = [props.size, props.variant, props.color, props.fullWidth ? 'block' : '']
  return styles
})
</script>
