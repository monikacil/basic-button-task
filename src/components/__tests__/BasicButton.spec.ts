import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import BasicButton from '../BasicButton.vue'

describe('BasicButton', () => {
  it('renders properly', () => {
    const wrapper = mount(BasicButton, {
      props: { color: 'red' },
      slots: {
        default: 'Button test'
      }
    })
    expect(wrapper.html()).toContain('Button test')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
