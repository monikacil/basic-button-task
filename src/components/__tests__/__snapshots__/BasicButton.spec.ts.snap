// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicButton > renders properly 1`] = `"<button class="btn m filled red">Button test</button>"`;
